src/renderer/components/swap/EditableAddress.tsx

```diff
@@ -0,0 +1,104 @@
+  asset: Asset
+  address: Address
+  onClickOpenAddress: (address: Address) => void
+  onChangeAddress: (address: Address) => void
+  addressValidator: AddressValidation
+}
+export const EditableAddress = ({
+  address,
+  onChangeAddress,
+  onClickOpenAddress,
+  addressValidator
+}: EditableAddressProps) => {
+  const RECIPIENT_FIELD = 'recipient'
+  const intl = useIntl()
+  const [editableAddress, setEditableAddress] = useState<O.Option<Address>>(O.none)
+  const maskedRecipientAddress = useMemo(() => address.substring(0, 7) + '...' + address.slice(-3), [address])
+
+  const validateAddress = useCallback(
+    async (_: unknown, value: string) => {
+      if (!value) {
+        return Promise.reject(intl.formatMessage({ id: 'wallet.errors.address.empty' }))
+      }
+      if (!addressValidator(value)) {
+        return Promise.reject(intl.formatMessage({ id: 'wallet.errors.address.invalid' }))
+      }
+    },
+    [addressValidator, intl]
+  )
+
+  const [form] = Form.useForm<{ recipient: string }>()
+
+  const renderCustomAddress = useCallback(() => {
+    return (
+      <>
+        <Styled.AddressCustomRecipient>
+          {maskedRecipientAddress}
+          <div>
+            <Styled.EditAddressIcon onClick={() => setEditableAddress(O.fromNullable(address))} />
+            <Styled.CopyLabel copyable={{ text: address }} />
+            <Styled.AddressLinkIcon onClick={() => onClickOpenAddress(address)} />
+          </div>
+        </Styled.AddressCustomRecipient>
+      </>
+    )
+  }, [address, maskedRecipientAddress, onClickOpenAddress])
+
+  const renderEditableCustomAddress = useCallback(
+    (editableAddress: Address) => {
+      return (
+        <Styled.EditableFormWrapper>
+          <Styled.AddressFormWrapper>
+            <InnerForm
+              form={form}
+              initialValues={{
+                recipient: editableAddress
+              }}>
+              <Form.Item rules={[{ required: true, validator: validateAddress }]} name={RECIPIENT_FIELD}>
+                <Input />
+              </Form.Item>
+            </InnerForm>
+          </Styled.AddressFormWrapper>
+          <Styled.AddressEditButtonsWrapper>
+            <Styled.ConfirmEdit
+              onClick={() => {
+                if (form.getFieldError(RECIPIENT_FIELD).length === 0) {
+                  onChangeAddress(form.getFieldValue(RECIPIENT_FIELD))
+                  setEditableAddress(O.fromNullable(null))
+                }
+              }}
+            />
+            <Styled.CancelEdit onClick={() => setEditableAddress(O.fromNullable(null))} />
+          </Styled.AddressEditButtonsWrapper>
+        </Styled.EditableFormWrapper>
+      )
+    },
+    [form, onChangeAddress, validateAddress]
+  )
+
+  const renderCustomAddressInput = useMemo(
+    () =>
+      FP.pipe(
+        editableAddress,
+        O.map((editableAddress) => renderEditableCustomAddress(editableAddress)),
+        O.getOrElse(() => renderCustomAddress())
```

Comment:
In some cases it might be more readable to use `O.fold` (kinda pattern matching) instead of `O.map ... O.getOrElse`

```ts
FP.pipe(
  editableAddress,
  O.fold(
    // in case of `none`
    () => renderCustomAddress(),
    // in case of `some(...)`
    renderEditableCustomAddress // similar to ((editableAddress) => renderEditableCustomAddress(editableAddress)
  )   

```